src/nn/tensor_collection/collection.rs

```diff
@@ -1,71 +1,156 @@
+pub trait ModuleVisitor<
+    T: TensorCollection<E, D>,
+    E: Dtype,
+    D: Device<E>,
+    E2: Dtype,
+    D2: Device<E2>,
+>: Sized
+{
     type Err;
+    type Func: TensorVisitor<E, D, E2, D2>;
 
-    /// Visit a [TensorCollection]
+    /// Visit a [TensorCollection]. Do not use this; use visit_fields instead.
     fn visit_module<Field, GetRef, GetMut>(
         &mut self,
         name: &str,
         get_refs: GetRef,
         get_muts: GetMut,
-    ) -> Result<(), Self::Err>
+    ) -> Result<Option<Field::Output<E2, D2>>, Self::Err>
     where
         GetRef: FnMut(&T) -> &Field,
         GetMut: FnMut(&mut T) -> &mut Field,
         Field: TensorCollection<E, D>;
 
-    /// Visits an actual named [Tensor]
+    /// Visits an actual named [Tensor]. Do not use this; use visit_fields instead.
     fn visit_tensor<S: Shape, GetRef, GetMut>(
         &mut self,
         name: &str,
         get_refs: GetRef,
         get_muts: GetMut,
         opts: TensorOptions<S, E, D>,
-    ) -> Result<(), Self::Err>
+    ) -> Result<Option<Tensor<S, E2, D2>>, Self::Err>
     where
         GetRef: FnMut(&T) -> &Tensor<S, E, D>,
         GetMut: FnMut(&mut T) -> &mut Tensor<S, E, D>;
+
+    /// Takes something that implements [ModuleFields] and function that takes
+    /// [ModuleFields::Output] and returns an instance of T.
+    fn visit_fields<M: ModuleFields<T, E, D>>(
+        &mut self,
+        fields: M,
+        builder: impl FnOnce(M::Output<E2, D2>) -> T::Output<E2, D2>,
+    ) -> Result<Option<T::Output<E2, D2>>, Self::Err> {
+        let options = fields.visit_fields(self)?;
+        Ok(M::handle_options(options).map(builder))
+    }
 }
 
-impl<S: Shape, E: Dtype, D: DeviceStorage> TensorCollection<E, D> for Tensor<S, E, D> {
-    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(visitor: &mut V) -> Result<(), V::Err> {
+impl<S: Shape, E: Dtype, D: Device<E>> TensorCollection<E, D> for Tensor<S, E, D> {
+    type Output<E2: Dtype, D2: Device<E2>> = Tensor<S, E2, D2>;
+
+    fn iter_tensors<E2: Dtype, D2: Device<E2>, V: ModuleVisitor<Self, E, D, E2, D2>>(
+        visitor: &mut V,
+    ) -> Result<Option<Self::Output<E2, D2>>, V::Err> {
         visitor.visit_tensor(
             "",
             |s| s,
             |s| s,
             TensorOptions {
                 do_gradient_update: true,
                 reset: |_| Ok(()),
+                shape: Default::default(),
             },
         )
     }
 }
 
 /// Options to change behavior of [ModuleVisitor]
 #[non_exhaustive]
-pub struct TensorOptions<S: Shape, E: Dtype, D: DeviceStorage> {
+pub struct TensorOptions<S: Shape, E: Dtype, D: Device<E>> {
     /// Whether the tensor should be updated with gradients
     pub do_gradient_update: bool,
 
     /// How to reset the tensor in the future.
     pub reset: fn(&'_ mut Tensor<S, E, D>) -> Result<(), D::Err>,
+
+    /// The [Shape] that BuildModule uses to construct the tensor
+    pub shape: S,
```

Comment:
How about Option<S>? Defaults of usize dims will be 0 which may lead to some subtle bugs

Reply:
I don't know. I'd like to keep it so that we don't have to specify the shape of tensors with ConstShapes, but I don't see a way to do this without trait specialization.